controller/common/common.go

```diff
@@ -64,28 +65,49 @@ func (m ChildMap) FindGroupKindName(apiGroup, kind, name string) *unstructured.U
 	return nil
 }
 
-func (m ChildMap) ReplaceChild(child *unstructured.Unstructured) {
+func (m ChildMap) ReplaceChild(prefix string, child *unstructured.Unstructured) {
 	key := childMapKey(child.GetAPIVersion(), child.GetKind())
 	children := m[key]
 	if children == nil {
 		// We only want to replace if it already exists, so do nothing.
 		return
 	}
 	name := child.GetName()
-	if _, found := children[name]; found {
-		children[name] = child
+	if _, found := children[prefix + name]; found {
+		children[prefix + name] = child
 	}
 }
 
-func MakeChildMap(list []*unstructured.Unstructured) ChildMap {
+// MakeChildMap builds the map of children resources that is suitable for use
+// in the `children` field of a CompositeController SyncRequest or
+// `attachments` field of  the  DecoratorControllers SyncRequest.
+//
+// This function returns a ChildMap which is a map of maps. The outer most map
+// is keyed  using the child's type and the inner map is keyed using the
+// child's name. If the parent resources is clustered and the child resource
+// is namespaced the inner map's keys are  prefixed by the namespace of the
+// child resource.
+//
+// This function requires parent resources has the meta.Namespace accurately
+// set. If the namespace of the pareent is empty it's considered a clustered
+// resource.
+//
+// If a user returns a namespaced as a child of a clustered resources without
+// the namespace set this is considered a user error but it's not handled here
+// since the api errorstrying to create the child is clear.
+func MakeChildMap(parent metav1.Object, list []*unstructured.Unstructured) ChildMap {
 	children := make(ChildMap)
 	for _, child := range list {
 		key := childMapKey(child.GetAPIVersion(), child.GetKind())
 
 		if children[key] == nil {
 			children[key] = make(map[string]*unstructured.Unstructured)
 		}
-		children[key][child.GetName()] = child
+		if parent.GetNamespace() == "" && child.GetNamespace() != "" {
```

Comment:
We don't actually call Insert and Replace here, in MakeChild map, it would probably make sense to put this logic in one place and call Insert and Replace from here.

Sorry it's been so long that this morning I actually missed that I had added prefix to both Insert and Replace. I think moving the logic into those functions makes more sense that adding a prefix, I'll update it.